### PR TITLE
BUG: Make ``arr = np.fromstring("\n", sep=" ")`` return ``[]``.

### LINKED ISSUE #18435: np.fromstring("\n", sep=" ") returns non-empty array
```python
import numpy as np

arr = np.fromstring("\n", sep=" ")
print(arr)
```
```
array([-1.]
```

### PR DESCRIPTION
**Pull request by tdakic**
Closes #18435.

### CODE REVIEW

File: numpy/core/src/multiarray/ctors.c [hunk 1]
```diff
@@ -73,7 +73,7 @@ fromstr_next_element(char **s, void *dptr, PyArray_Descr *dtype,
     int r = dtype->f->fromstr(*s, dptr, &e, dtype);
     /*
      * fromstr always returns 0 for basic dtypes; s points to the end of the
-     * parsed string. If s is not changed an error occurred or the end was
+     * parsed string. If s is not changed an error occu5trred or the end was
      * reached.
```
Comment (charris) [inline numpy/core/src/multiarray/ctors.c:77]: Spelling.

Comment (charris) [reply numpy/core/src/multiarray/ctors.c:77]: Could also use a comma after "changed".

**Review by tdakic — COMMENTED** [review]
Yes, a typo :-(

**Review by tdakic — COMMENTED** [review]
Fixed!

**Comment by tdakic** [comment]
With this fix, `print(np.fromstring(", 1,2,3", sep= ",",dtype =np.int64))` will print `[1 2 3]`. Should it?

**Comment by charris** [comment]
close/reopen

**Comment by charris** [comment]
Looking at this, I'm not convinced the result should be `[]`. Apparently white space (`\n`) is interpreted as `-1`. For instance
```
In [17]: fromstring(' , ',sep=',')                                              
Out[17]: array([-1., -1.])
```
That may be a problem, but it is a different problem. I'm not yet clear where that value comes from. @seberg Thoughts?

**Comment by seberg** [comment]
Uh, that is strange... It seems you are right and whitespace is parsed as a number and that number is `-1.`... That explains how the `-1.` can be so persistent.

It basically drops out of this:
https://github.com/numpy/numpy/blob/6ee49178517088966e63c2aedf6a8a5779ad5384/numpy/core/src/common/numpyos.c#L429-L450

Now, that should indicate that an error occurred, by resetting the end-pointer to the start position on error, which then should trigger the error return path here:
https://github.com/numpy/numpy/blob/bb7a31a6a852ecc64d3f7cffb70121fc8bef20eb/numpy/core/src/multiarray/ctors.c#L72-L79

Which then has a negative return value indicating an error in the read.

All of that seems a bit strange and every single level of that code looks like it could be smoothened a bit more.  Maybe we should try and see if we can do a proper deprecation of that `-1` if the `-1` works for reliably for everything that cannot be read.

The bug seems to be that call to `string_is_fully_read` which probably does the right thing for the `skip` function, but not for the parse function. Since it checks `start >= end` and equality holds here.

The easier fix should probably to properly propagate a `-1` return value when not successfully reading from `dtype->f->fromstr` (or even lower).  The only question is whether we want to try a deprecation there, I am not sure how easy it is.

Long story short, the `-1.` is definitely an accident (I wonder if it has always been there, or an artifact of fixing another bug).  But if we want to gently "deprecate" it, I guess it might be a bit tricky.  If we are happy to not be gentle about it, I expect the `string_is_fully_read` might just be unnecessary after an unsuccessfull value parsing attempt.  (worst case, we have to check for a leading `\0` maybe, in case the string is fully empty.

**Comment by jorenham** [comment]
This seems to have stranded, so I'm going to close it. If you want to continue working on this, @tdakic, let me know, and we'll re-open it.
